File: surf-api-bukkit/surf-api-bukkit-server/src/main/java/dev/slne/surf/surfapi/bukkit/server/impl/nms/bridges/packets/player/SurfBukkitNmsPlayerPacketsImpl.java
package dev.slne.surf.surfapi.bukkit.server.impl.nms.bridges.packets.player;

import static com.google.common.base.Preconditions.checkNotNull;

import dev.slne.surf.surfapi.bukkit.api.nms.bridges.packets.PacketOperation;
import dev.slne.surf.surfapi.bukkit.api.nms.bridges.packets.player.SurfBukkitNmsPlayerPackets;
import dev.slne.surf.surfapi.bukkit.server.impl.nms.bridges.packets.PacketOperationImpl;
import dev.slne.surf.surfapi.bukkit.server.nms.NmsUtil;
import io.papermc.paper.adventure.PaperAdventure;
import io.papermc.paper.math.BlockPosition;
import javax.annotation.ParametersAreNonnullByDefault;
import net.kyori.adventure.text.Component;
import net.minecraft.network.protocol.game.ClientboundBlockUpdatePacket;
import net.minecraft.network.protocol.game.ClientboundContainerClosePacket;
import net.minecraft.network.protocol.game.ClientboundContainerSetSlotPacket;
import net.minecraft.network.protocol.game.ClientboundOpenScreenPacket;
import net.minecraft.network.protocol.game.ClientboundOpenSignEditorPacket;
import org.bukkit.event.inventory.InventoryType;
import org.bukkit.inventory.ItemStack;

@ParametersAreNonnullByDefault
public final class SurfBukkitNmsPlayerPacketsImpl implements SurfBukkitNmsPlayerPackets, NmsUtil {

  @Override
  public PacketOperation openSignEditor(BlockPosition position, boolean frontSide) {
    checkNotNull(position, "position");

    return PacketOperationImpl.simple(
        player -> new ClientboundOpenSignEditorPacket(toNms(position), frontSide));
  }

  @Override
  public PacketOperation openInventory(int syncId, InventoryType type, Component title) {
    checkNotNull(type, "type");
    checkNotNull(title, "title");
    return PacketOperationImpl.simple(player -> new ClientboundOpenScreenPacket(syncId, toNms(type),
        PaperAdventure.asVanilla(title)));
  }

  @Override
  public PacketOperation setInventorySlot(int syncId, int revision, int slot, ItemStack item) {
    checkNotNull(item, "item");

    return PacketOperationImpl.simple(
        player -> new ClientboundContainerSetSlotPacket(syncId, revision, slot, toNms(item)));
  }

  @Override
  public PacketOperation closeInventory(int syncId) {
    return PacketOperationImpl.simple(player -> new ClientboundContainerClosePacket(syncId));
  }
}

File: surf-api-bukkit/surf-api-bukkit-server/src/main/kotlin/dev/slne/surf/surfapi/bukkit/server/impl/nms/bridges/packets/player/SurfBukkitNmsPlayerPacketsImpl.kt
package dev.slne.surf.surfapi.bukkit.server.impl.nms.bridges.packets.player

import com.google.auto.service.AutoService
import dev.slne.surf.surfapi.bukkit.api.nms.NmsUseWithCaution
import dev.slne.surf.surfapi.bukkit.api.nms.bridges.packets.player.SurfBukkitNmsPlayerPackets
import dev.slne.surf.surfapi.bukkit.server.impl.nms.bridges.packets.PacketOperationImpl
import dev.slne.surf.surfapi.bukkit.server.nms.toNms
import dev.slne.surf.surfapi.core.api.util.checkInstantiationByServiceLoader
import io.papermc.paper.math.BlockPosition
import net.kyori.adventure.text.Component
import net.minecraft.network.protocol.game.ClientboundContainerClosePacket
import net.minecraft.network.protocol.game.ClientboundContainerSetSlotPacket
import net.minecraft.network.protocol.game.ClientboundOpenScreenPacket
import net.minecraft.network.protocol.game.ClientboundOpenSignEditorPacket
import org.bukkit.event.inventory.InventoryType
import org.bukkit.inventory.ItemStack

@AutoService(SurfBukkitNmsPlayerPackets::class)
@NmsUseWithCaution
class SurfBukkitNmsPlayerPacketsImpl : SurfBukkitNmsPlayerPackets {
    init {
        checkInstantiationByServiceLoader()
    }

    override fun openSignEditor(
        position: BlockPosition,
        frontSide: Boolean,
    ) = PacketOperationImpl.simple { ClientboundOpenSignEditorPacket(position.toNms(), frontSide) }

    override fun openInventory(
        syncId: Int,
        type: InventoryType,
        title: Component,
    ) = PacketOperationImpl.simple {
        ClientboundOpenScreenPacket(
            syncId,
            type.toNms(),
            title.toNms()
        )
    }

    override fun setInventorySlot(
        syncId: Int,
        revision: Int,
        slot: Int,
        item: ItemStack,
    ) = PacketOperationImpl.simple {
        ClientboundContainerSetSlotPacket(
            syncId,
            revision,
            slot,
            item.toNms()
        )
    }

    override fun closeInventory(syncId: Int) =
        PacketOperationImpl.simple { ClientboundContainerClosePacket(syncId) }
}
